src/System.Diagnostics.Process/src/System/Diagnostics/Process.NonUap.cs

```diff
@@ -0,0 +1,115 @@
+                EnsureState(State.Associated | State.IsLocal | State.HaveNonExitedId);
+
+                if (IsSelfOrDescendant(GetCurrentProcess()))
+                    throw new InvalidOperationException(SR.KillEntireProcessTree_DisallowedBecauseTreeContainsCallingProcess);
+
+                IEnumerable<Exception> result = KillTree();
+
+                if (result.Any())
+                    throw new AggregateException(SR.KillEntireProcessTree_TerminationIncomplete, result);
+            }
+        }
+
+        private bool IsSelfOrDescendant(Process processOfInterest)
+        {
+            if (SafePredicateTest(() => Equals(processOfInterest)))
+                return true;
+
+            Process[] allProcesses = GetProcesses();
+
+            try
+            {
+                var descendantProcesses = new Queue<Process>();
+                Process current = this;
+
+                do
+                {
+                    foreach (Process candidate in current.GetChildProcesses(allProcesses))
+                    {
+                        if (SafePredicateTest(() => processOfInterest.Equals(candidate)))
+                            return true;
+
+                        descendantProcesses.Enqueue(candidate);
+                    }
+                } while (descendantProcesses.TryDequeue(out current));
+            }
+            finally
+            {
+                foreach (Process process in allProcesses)
+                {
+                    process.Dispose();
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Returns all immediate child processes.
+        /// </summary>
+        private IReadOnlyList<Process> GetChildProcesses(Process[] processes = null)
+        {
+            bool internallyInitializedProcesses = processes == null;
+            processes = processes ?? GetProcesses();
+
+            List<Process> childProcesses = new List<Process>();
+
+            foreach (Process possibleChildProcess in processes)
+            {
+                // Only support disposing if this method initialized the set of processes being searched
+                bool dispose = internallyInitializedProcesses;
+
+                try
+                {
+                    if (SafePredicateTest(() => IsParentOf(possibleChildProcess)))
+                    {
+                        childProcesses.Add(possibleChildProcess);
+                        dispose = false;
+                    }
+                }
+                finally
+                {
+                    if (dispose)
+                        possibleChildProcess.Dispose();
+                }
+            }
+
+            return childProcesses;
+        }
+
+        private bool SafePredicateTest(Func<bool> predicate)
+        {
+            try
+            {
+                return predicate();
+            }
+            catch (Exception e) when (e is InvalidOperationException || e is Win32Exception)
```

Comment:
This is a very broad try catch.
An alternative would be to:
- not catch Win32Exceptions, e.g. permission issues should bubble up
- make ParentProcessId a nullable int, that is null when the process has exited.

Reply:
It is also a bit friendlier for debugging if we try to not throw 1st chance exceptions in a "fairly normal" case like "already exited"

Reply:
This bit of code is used only when determining which processes are children, not when performing process termination (so it's not suppressing exceptions encountered when attempting termination).

On the Windows side, a process doesn't necessarily have permission to access details on all other processes so encountering permissions errors can be a part of the normal course of business (e.g. a non-elevated process's attempting to get details on an elevated process may fail with a Win32Exception). In this context, if our goal is to make a best-effort attempt to terminate a process tree, it seems like these exceptions should be suppressed.

Reply:
makes sense